modules/native_streaming_client_module/src/native_streaming_impl.cpp

```diff
@@ -93,109 +106,189 @@ void NativeStreamingImpl::connectionStatusChangedHandler(opendaq_native_streamin
+{
+    if (status == ClientConnectionStatus::Connected)
+    {
+        this->transportClientHandler->sendStreamingRequest();
+        protocolInitPromise = std::promise<void>();
+        protocolInitFuture = protocolInitPromise.get_future();
+        protocolInitTimer->expires_from_now(streamingInitTimeout);
+        protocolInitTimer->async_wait(
+            [this](const boost::system::error_code& ec)
+            {
+                if (ec)
+                    return;
+
+                if (protocolInitFuture.wait_for(std::chrono::milliseconds(0)) == std::future_status::ready)
+                    updateConnectionStatus(ClientConnectionStatus::Connected);
+                else
+                    updateConnectionStatus(ClientConnectionStatus::Unrecoverable);
+            }
+        );
+    }
+    else
+    {
+        updateConnectionStatus(status);
+    }
+}
+
+void NativeStreamingImpl::upgradeClientHandlerCallbacks()
 {
     using namespace boost::asio;
+    WeakRefPtr<IStreaming> thisRef = this->template borrowPtr<StreamingPtr>();
 
     OnSignalAvailableCallback signalAvailableCb =
-        [this](const StringPtr& signalStringId,
+        [this, thisRef](const StringPtr& signalStringId,
                const StringPtr& serializedSignal)
     {
         dispatch(
             *processingIOContextPtr,
-            processingStrand.wrap(
-                [this, signalStringId = signalStringId, serializedSignal = serializedSignal]()
-                {
+            [this, thisRef, signalStringId = signalStringId, serializedSignal = serializedSignal]()
+            {
+                if (auto thisPtr = thisRef.getRef(); thisPtr.assigned())
                     signalAvailableHandler(signalStringId, serializedSignal);
-                }
-            )
+            }
         );
     };
     OnSignalUnavailableCallback signalUnavailableCb =
-        [this](const StringPtr& signalStringId)
+        [this, thisRef](const StringPtr& signalStringId)
     {
         dispatch(
             *processingIOContextPtr,
-            processingStrand.wrap(
-                [this, signalStringId = signalStringId]()
-                {
+            [this, thisRef, signalStringId = signalStringId]()
+            {
+                if (auto thisPtr = thisRef.getRef(); thisPtr.assigned())
                     signalUnavailableHandler(signalStringId);
-                }
-            )
+            }
         );
     };
     OnPacketCallback onPacketCallback =
-        [this](const StringPtr& signalStringId, const PacketPtr& packet)
+        [this, thisRef](const StringPtr& signalStringId, const PacketPtr& packet)
     {
         dispatch(
             *processingIOContextPtr,
-            processingStrand.wrap(
-                [this, signalStringId = signalStringId, packet = packet]()
-                {
+            [this, thisRef, signalStringId = signalStringId, packet = packet]()
```

Comment:
Why capture `signalStringId = signalStringId` and `packet = packet`? Why not just `signalStringId ` and `packet`?